rtp.1
.\" Copyright (c) 2018 Jan Stary <hans@stare.cz>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.Dd June 29, 2018
.Dt RTP 1
.Os
.Sh NAME
.Nm rtp
.Nd debug RTP sessions
.Sh SYNOPSIS
.Nm
.Op Fl r
.Op Fl t
.Op Fl v
.Op Fl i Ar format
.Op Fl o Ar format
.Op input
.Op output
.Sh DESCRIPTION
.Nm
reads a stream of RTP packets from
.Ar input
and writes it to
.Ar output .
Both can be a named file or a network address.
By default,
.Nm
reads from standard input and writes to standard output.
A name of
.Sq -
also means standard input or output.
A network address is given as
.Oo Ar address Oc Ns Ar :port ,
where
.Ar address
is either an IP address in dot-decimal notation or a hostname
.Po
.Dq localhost
by default
.Pc ,
and
.Ar port
is a number.
.Pp
If the
.Ar output
is a local address
(i.e. an address assigned to a network interface of the local machine),
.Nm
will wait till someone sends a one-byte message to that address
and then will stream the input to them.
Similarly, if the
.Ar input
is a remote address,
.Nm
will try to send a one-byte message first to let them know we are reading.
For other combinations of input and output,
.Nm
behaves in the obvious way.
.Pp
The stream can be read or written in the following formats:
.Bl -tag -width Ds
.It Cm dump
The format of the original rtptools (see below).
A dump file starts with a text line of the form
.Dq #!rtptools1.0 addr/port ,
followed by a dumpfile header:
.Bd -literal
struct dumphdr {
	struct {
		uint32_t sec;
		uint32_t usec;
	}		time;
	uint32_t	addr;
	uint16_t	port;
	uint16_t	zero;
};
.Ed
.Pp
When reading,
.Nm
checks that the
.Va addr
and
.Va port
specified at the starting line are those stored in the above header.
When writing,
.Nm
sets them to the address and port where the traffic was captured.
.Pp
The actual RTP packets follow, each prefixed with:
.Bd -literal
struct dpkthdr {
	uint16_t dlen; /* stored packet length, including this   */
	uint16_t plen; /* original RTP (header + payload) length */
	uint32_t usec; /* dump time: usec since start time       */
};
.Ed
.Pp
If the payload was truncated during dump (see the
.Fl x
option),
.\"FIXME
dlen will be shorter than plen.
The RTP header is always there.
All fields are stored in the network byte order.
.It Cm pcap
The ubiquituous format of
.Xr libpcap 3 .
This format can only be used as an input; a
.Cm pcap
output can be created with
.Xr tcpdump 1
by sniffing an interface where the RTP traffic happens.
The main difference to the
.Cm dump
format is that a
.Cm pcap
input can contain more than one RTP stream.
The output must be to a file;
.Nm
will create one output file for each stream found in the input.
.\" FIXME suitably named
.It Cm net
The actual RTP packets being sent and received.
This is the only format used with network connections.
.It Cm raw
The actual audio payload of the RTP packet.
This format can only be used as an output to a file.
.Nm
has no support for the actual audio
and will not deal with the audio codec involved.
.It Cm txt
RTP packets described with lines of text.
.El
.Pp
For regular files, the format will be guessed from the file name suffix:
.Dq rtp
for
.Cm dump ,
.Dq pcap
for
.Cm pcap ,
.Dq raw
for
.Cm raw ,
and
.Dq txt
for
.Cm txt ,
with
.Cm dump
being the default if the format cannot be guessed from the name.
.Pp
By default,
.Nm
will use the RTP timestamps when sending packets out.
This can be changed with the
.Fl t
option, which uses the dump time instead.
.Pp
The options are as follows.
.Pp
.Bl -tag -compact -width formatxxx
.It Fl i Ar format
Set the input format.
.It Fl o Ar format
Set the output format.
.It Fl r
Treat all addresses as remote.
.It Fl t
Use dump time for outgoing packets.
.It Fl v
Be verbose about the packets.
.El
.Sh EXAMPLES
Read from a dump file, save the raw audio:
.Pp
.Dl $ rtp session.rtp session.raw
.Pp
Read from a dump file, stream to whoever connects:
.Pp
.Dl $ rtp session.rtp localhost:1234
.Pp
Read from a dump file, stream to a remote location:
.Pp
.Dl $ rtp session.rtp far.away.com:1234
.Pp
Read rtp on a local port, save a textual description:
.Pp
.Dl $ rtp localhost:1234 outfile.txt
.Pp
Read a rtp stream from a remote location, stream it to somewhere else.
This makes
.Nm
a RTP tunnel.
.Pp
.Dl $ rtp far.away.com:1234 somewhere.else.com:3456
.Pp
Read a rtp stream from a remote location, write audio payload to stdout.
This makes
.Nm
an RTP radio listener.
.Pp
.Dl $ rtp -o raw radio.com:1234
.Dl $ rtp -o raw radio.com:1234 | play -t raw -c 1 -r 8000 -e u-law -
.Pp
With the
.Fl r
option, network functionality can be run locally.
For example, the following two instances cannot run simultaneously:
.Pp
.Dl $ rtp session.rtp localhost:1234
.Dl $ rtp localhost:1234 session.raw
.Pp
As
.Dq localhost
resolves to a local address, the first instance will
.Xr bind 2
to port 1234 and stream the file to whoever connects, as described above.
The second instance will also try to
.Xr bind 2
to the port, and store whatever arrives there to the output file.
The second
.Xr bind 2
will fail.
Adding the
.Fl r
option to the second instance solves this
(whichever of the two is the second, in fact): in
.Pp
.Dl $ rtp    session.rtp localhost:1234
.Dl $ rtp -r localhost:1234 session.raw
.Pp
the second instance will treat
.Dq localhost
as a remote address, so the second
.Nm
will just connect there and read the output of the first one as its input.
Similarly, with
.Pp
.Dl $ rtp    localhost:1234 session.raw
.Dl $ rtp -r session.rtp localhost:1234
.Pp
the first one will wait for an input connection,
and the second will make that connection, treating
.Dq localhost
as a remote address.
.Pp
The following will set up a tunnel, a reader, and a writer:
.Pp
.Dl $ rtp    127.0.0.1:1234 127.0.0.1:3456
.Dl $ rtp -r 127.0.0.1:3456 session.raw
.Dl $ rtp -r session.rtp 127.0.0.1:1234
.Pp
The first instance (the tunnel) will wait for a connection to port 3456,
and then will read whatever arrives at 1234 and stream it out.
The second instance (the reader) will connect to 3456
(treating 127.0.0.1 as a remote address), waiting to read.
The third instance (the writer) will connect to 1234
(treating 127.0.0.1 as a remote address),
and will stream the content of the dump file there,
starting the pipeline flow.
.Sh HISTORY
In the early days of RTP, Henning Schulzrinne wrote a set of
.Dq rtptools
to debug RTP sessions;
.Nm
aims to replace rtptools.
Reading a local port and dumping it to a file is rtpdump.
Streaming a file to a remote location is rtpplay and rtpsend.
Reading from an address and streaming to another address is rtptrans.
.Sh AUTHORS
.An Jan Starý Aq Mt hans@stare.cz
.Sh BUGS
By convention, RTP traffic happens on an even port number,
and the corresponding RTCP traffic happens on the odd port+1.
.Nm
ignores that convention, only reads the specified
.Ar port ,
and misses the RTCP packets.


rtp.c
/*
 * Copyright (c) 2018 Jan Stary <hans@stare.cz>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <ctype.h>
#include <stdio.h>
#include <errno.h>
#include <err.h>

#include <sys/time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <ifaddrs.h>
#include <netdb.h>

#include "format-dump.h"
#include "format-rtp.h"

#define BUFLEN 8192
/* FIXME: This should be enough for each and every packet we read,
 * but we are still wrong: mind the buflen in the reading routines. */

extern const char* __progname;
struct ifaddrs *ifaces = NULL;
struct sockaddr_in *addr;

typedef enum {
	FORMAT_DUMP,
	FORMAT_NET,
	FORMAT_RAW,
	FORMAT_TXT,
	FORMAT_NONE
} format_t;

struct format {
	format_t	type;
	const char*	name;
	const char*	suff;
} formats[] = {
	{ FORMAT_DUMP,	"dump",	"rtp"	},
	{ FORMAT_NET,	"net",	NULL	},
	{ FORMAT_RAW,	"raw",	"raw"	},
	{ FORMAT_TXT,	"txt",	"txt"	},
	{ FORMAT_NONE,	NULL,	NULL	}
};
#define NUMFORMATS (sizeof(formats) / sizeof(struct format))

struct pt {
	const char*	enc;  /* encoding name */
	uint32_t	rate; /* sampling rate (audio) or clock rate (video) */
	uint8_t		ch;   /* audio channels; 0 for video */
} payload[] = {
	{ "PCMU",	 8000,	1 },
	{ "reserved",	    0,	0 },
	{ "reserved",	    0,	0 },
	{ "GSM ",	 8000,	1 },
	{ "G723",	 8000,	1 },
	{ "DVI4",	 8000,	1 },
	{ "DVI4",	16000,	1 },
	{ "LPC ",	 8000,	1 },
	{ "PCMA",	 8000,	1 },
	{ "G722",	 8000,	1 },
	{ "L16 ",	44100,	2 },
	{ "L16 ",	44100,	1 },
	{ "QCELP",	 8000,	1 },
	{ "CN  ",	 8000,	0 },
	{ "MPA ",	90000,	0 },
	{ "G728",	 8000,	1 },
	{ "DVI4",	11025,	1 },
	{ "DVI4",	22050,	1 },
	{ "G729",	 8000,	1 },
	{ "reserved",	    0,	0 },
	{ "unassigned",	    0,	0 },
	{ "unassigned",	    0,	0 },
	{ "unassigned",	    0,	0 },
	{ "unassigned",	    0,	0 },
	{ "unassigned",	    0,	0 },
	{ "CelB",	90000,	0 },
	{ "JPEG",	90000,	0 },
	{ "unassigned",	    0,	0 },
	{ "nv  ",	90000,	0 },
	{ "unassigned",	    0,	0 },
	{ "unassigned",	    0,	0 },
	{ "H261",	90000,	0 },
	{ "MPV ",	90000,	0 },
	{ "MP2T",	90000,	0 },
	{ "H263",	90000,	0 },
	{ NULL,		    0,	0 }
};
#define NUMPAYLOAD ((sizeof(payload)) / (sizeof(struct pt)))

static int remote = 0;
static int dumptime = 0;
static int verbose = 0;
static format_t ifmt = FORMAT_NONE;
static format_t ofmt = FORMAT_NONE;

static void
usage(void)
{
	fprintf(stderr,
		"%s [-r] [-v] [-i format] [-o format] [input] [output]\n",
		__progname);
}

format_t
fmtbyname(const char *name)
{
	struct format *fmt;
	for (fmt = formats; fmt && fmt->name; fmt++)
		if (strcmp(fmt->name, name) == 0)
			return fmt->type;
	return FORMAT_NONE;
}

format_t
fmtbysuff(const char *suff)
{
	struct format *fmt;
	for (fmt = formats; fmt && fmt->name; fmt++)
		if (fmt->suff && strcmp(fmt->suff, suff) == 0)
			return fmt->type;
	return FORMAT_NONE;
}

int
islocal(struct sockaddr_in *a)
{
	struct ifaddrs *i;
	if (remote)
		return 0;
	for (i = ifaces; i; i = i->ifa_next)
         	if (a->sin_addr.s_addr ==
		((struct sockaddr_in*)i->ifa_addr)->sin_addr.s_addr)
			return 1;
	return 0;
}

/* Open a path for reading or writing (the flags say which).
 * Set the input/output format, set addr/port if applicable.
 * Return a file descriptor, or -1 for failure. */
int
rtpopen(const char *path, int flags)
{
	int e;
	int fd = -1;
	uint16_t port;
	const char* er;
	char* p = NULL;
	format_t fmt = FORMAT_NONE;;
	struct addrinfo *res = NULL;
	struct addrinfo info;
	if (strcmp(path, "-") == 0) {
		/* stdin/stdout */
		if (flags & O_CREAT) {
			if (ofmt == FORMAT_NONE)
				ofmt = FORMAT_TXT;
			return STDOUT_FILENO;
		} else {
			if (ifmt == FORMAT_NONE)
				ifmt = FORMAT_DUMP;
			return STDIN_FILENO;
		}
	} else if ((p = strchr(path, ':'))) {
		/* addr:port */
		*p++ = '\0';
		if ((port = strtonum(p, 1, UINT16_MAX, &er)) == 0) {
			warnx("port number '%s' %s", p, er);
			return -1;
		}
		if (*path == '\0')
			path = NULL;
		memset(&info, 0, sizeof(info));
		info.ai_family = PF_INET;
		info.ai_socktype = SOCK_DGRAM;
		info.ai_protocol = IPPROTO_UDP;
		info.ai_flags = AI_ADDRCONFIG | AI_NUMERICSERV;
		if ((e = getaddrinfo(path, p, &info, &res))) {
			warnx("'%s': %s", path, gai_strerror(e));
			return -1;
		}
		if (res == NULL) {
			warn("getaddrinfo %s", path);
			goto bad;
		}
		if (-1 == (fd =
		socket(res->ai_family, res->ai_socktype, res->ai_protocol))) {
			warn("socket");
			goto bad;
		}
		if (-1 == setsockopt(fd,
		SOL_SOCKET, SO_REUSEADDR, &fd, sizeof(fd)))
			warn("REUSEADDR");
		/* TODO: SO_SNDTIMEO SO_RCVTIMEO SO_TIMESTAMP */
		addr = (struct sockaddr_in*) res->ai_addr;
		addr->sin_port = port;
		if (islocal(addr)) {
			/* If the local socket is an input, we will read on it;
			 * if it's an output, we want to receive a message first
			 * to know who to write to. So bind(2) in any case. */
			if (bind(fd, res->ai_addr, res->ai_addrlen) == -1) {
				warn("bind");
				goto bad;
			}
			if (flags & O_CREAT) {
				/* If this local address is an output,
				 * wait till we recvfrom() something first,
				 * to let us know where to send() later. */
				char b[2];
				socklen_t len;
				struct sockaddr r;
				while (recvfrom(fd, b, 1, 0, &r, &len) != 1)
					;
				if (connect(fd, &r, len) == -1) {
					/* Debian: EAFNOSUPPORT for localhost */
					warn("connect to output");
					goto bad;
				}
			}
		} else {
			if (connect(fd, res->ai_addr, res->ai_addrlen) == -1) {
				warn("connect to remote");
				goto bad;
			}
			if (!(flags & O_CREAT)) {
				/* If the remote address is our input,
			 	 * send a byte there to let them know
				 * where to send(2) later. */
				char b[2] = "1";
				if (send(fd, b, 1, 0) != 1) {
					warn("send");
					goto bad;
				}
			}
		}
		freeaddrinfo(res);
		if (flags & O_CREAT) {
			if (ofmt == FORMAT_NONE) {
				ofmt = FORMAT_NET;
			} else if (ofmt != FORMAT_NET) {
				warnx("Only net output allowed for %s", path);
				goto bad;
			}
		} else {
			if (ifmt == FORMAT_NONE) {
				ifmt = FORMAT_NET;
			} else if (ifmt != FORMAT_NET) {
				warnx("Only net input allowed for %s", path);
				goto bad;
			}
		}
	} else {
		if (-1 == (fd = (flags & O_CREAT)
		? open(path, flags, 0644)
		: open(path, flags))) {
			warn("%s", path);
			goto bad;
		}
		if ((p = strrchr(path, '.')))
			fmt = fmtbysuff(++p);
		/* Be compatible with rtptools */
		if ((flags & O_CREAT) && (ofmt == FORMAT_NONE)) {
			ofmt = (fmt != FORMAT_NONE) ? fmt : FORMAT_TXT;
		} else if (ifmt == FORMAT_NONE) {
			ifmt = (fmt != FORMAT_NONE) ? fmt : FORMAT_DUMP;
		}
	}
	return fd;
bad:
	if (res)
		freeaddrinfo(res);
	close(fd);
	return -1;
}

/* Compute the time interval from old to new timeval.
 * Return the difference in usec, or 0 for error.
 * This is used to time outgoing packets,
 * so the 0 means "send immediately". */
uint32_t
tvdiff(struct timeval *old, struct timeval *new)
{
	uint32_t usec;
	if (old == NULL || new == NULL)
		return 0;
	if ((old->tv_sec > new->tv_sec)
	|| ((old->tv_sec == new->tv_sec) && (old->tv_usec > new->tv_usec)))
		return 0;
	usec = (new->tv_sec - old->tv_sec) * 1000000;
	usec += new->tv_usec - old->tv_usec;
	return usec;
}

/* Compute the offset since the given timeval.
 * Return the difference in msec, or 0 for error. */
uint32_t
offset(struct timeval *since)
{
	struct timeval now;
	if (gettimeofday(&now, NULL) == -1) {
		warn("gettimeofday");
		return 0;
	}
	return tvdiff(since, &now) / 1000;
}

/* The 'zero' describes the start of the dump,
 * the 'when' says (in msec since zero) when the next packet goes out.
 * Sleep for the appropriate time; then return 0, or -1 if interrupted. */
int
dumpsleep(struct timeval *zero, uint32_t when)
{
	struct timeval now;
	struct timespec nap;
	uint32_t diff, usec = when * 1000;
	/* some time has already elapsed */
	if (gettimeofday(&now, NULL) == -1) {
		warnx("gettimeofday");
		return -1;
	}
	if ((diff = tvdiff(zero, &now)) > usec) {
		/* we are late already */
		return 0;
	}
	usec -= diff;
	nap.tv_sec = usec / 1000000;
	nap.tv_nsec = (usec % 1000000) * 1000;
	if (nanosleep(&nap, NULL) != 0) {
		warn("nanosleep");
		return -1;
	}
	return 0;
}

/* The meaning of the rtp timestamp is payload dependent.
 * Use the global paylod type table to compute the appropriate
 * time to sleep, where 'last' is the rtp timestamp of the last
 * rtp packet we sent, and 'next' is the timestamp of the next.
 * For example, a typical audio application uses a 8000 kHz rate
 * and the RTP timestamp increments by 160. This means a step of
 * 160/8000 = 1/50 of a second = 20 ms.
 * Return 0 for success, -1 for error. */
int
rtpsleep(uint32_t *last, uint32_t next, uint8_t pt)
{
	double step;
	uint32_t diff;
	struct timespec nap;
	if (*last == 0) {
		/* first packet */
		*last = next;
		return 0;
	}
	if (*last > next) {
		warnx("packets out of timestamp order: %u > %u", *last, next);
		return -1;
	}
	if (pt > NUMPAYLOAD) {
		warnx("unknown payload %u: sending immediately", pt);
		return -1;
	}
	if (payload[pt].rate == 0) {
		warnx("unknown rate for '%s': sending now", payload[pt].enc);
		return -1;
	}
	diff = next - *last; /* FIXME: minus time already elapsed. */
	*last = next;
	step = (1.0 * diff) / payload[pt].rate;
	nap.tv_sec = step;
	nap.tv_nsec = (step - nap.tv_sec) * 1000000000;
	if (nanosleep(&nap, NULL) != 0) {
		warn("nanosleep");
		return -1;
	}
	return 0;
}

/* Read a dump file from input, send the RTP output via net.
 * Return 0 for success, -1 for error. */
int
dump2net(int ifd, int ofd)
{
	ssize_t r, w;
	int error = 0;
	struct sockaddr_in addr;
	struct dumphdr hdr;
	struct dpkthdr *pkt;
	struct rtphdr  *rtp;
	struct timeval zero;
	uint32_t last = 0;
	unsigned char buf[BUFLEN];
	if (read_dumpline(ifd, &addr) == -1) {
		warnx("Error reading dump line");
		return -1;
	}
	if (read_dumphdr(ifd, &hdr, DUMPHDRSIZE) == -1) {
		warnx("Error reading %zd bytes of dump header", DUMPHDRSIZE);
		return -1;
	}
	if (check_dumphdr(&hdr, &addr) == -1)
		warnx("Dump file header is inconsistent");
	if (verbose)
		print_dumphdr(&hdr);
	if (dumptime && gettimeofday(&zero, NULL) == -1) {
		warnx("gettimeofday");
		return -1;
	}
	while ((r = read_dump(ifd, buf, BUFLEN)) > 0) {
		pkt = (struct dpkthdr*) buf;
		rtp = (struct rtphdr*) (buf + DPKTHDRSIZE);
		if (pkt->plen == 0) { /* FIXME: that's RTCP. Currently, we don't
			send these, because receiving zero size confuses the
			reader, who considers that an end. But a RTCP packet
			does not actualy have zero size. We need to properly
			read the RTCP header, which we don't, yet. */
			continue;
		}
		if ((dumptime
		? dumpsleep(&zero, pkt->msec)
		: rtpsleep(&last, ntohl(rtp->ts), rtp->pt)) == -1) {
		/* FIXME: notice how we use pkt->msec, because that's
		 * already converted to the local byte order by
		 * read_dump() -> read_dpkthdr(); but we convert the rtp->ts,
		 * because we have not properly parsed the RTP packet;
		 * which we have to, e.g. to change the timestamp. */
			warnx("packet timing failed");
			error = -1;
			continue;
		}
		if (verbose)
			print_dpkthdr(pkt);
		if (parse_rtphdr(rtp) == -1) {
			warnx("Error parsing RTP header");
			error = -1;
			continue;
		}
		if (verbose)
			print_rtphdr(rtp);
		if ((w = send(ofd, rtp, pkt->plen, 0)) == -1) {
			warnx("Error sending %u bytes of RTP", pkt->plen);
			error = -1;
			continue;
		} else if (w < pkt->plen) {
			warnx("Only sent %zd < %u bytes of RTP", w, pkt->plen);
			error = -1;
			continue;
		}
	}
	return r == -1 ? -1 : error;
}

/* Read a dump file from input, write raw audio payload to output.
 * Return 0 for success, -1 for error. */
int
dump2raw(int ifd, int ofd)
{
	struct sockaddr_in addr;
	struct dumphdr hdr;
	struct dpkthdr *pkt;
	struct rtphdr *rtp;
	unsigned char buf[BUFLEN];
	unsigned char *p = buf;
	ssize_t r, w, hlen;
	int error = 0;
	if (read_dumpline(ifd, &addr) == -1) {
		warnx("Error reading dump line");
		return -1;
	}
	if (read_dumphdr(ifd, &hdr, DUMPHDRSIZE) == -1) {
		warnx("Error reading dump header");
		return -1;
	}
	if (check_dumphdr(&hdr, &addr) == -1)
		warnx("Dump file header is inconsistent");
	if (verbose)
		print_dumphdr(&hdr);
	while ((r = read_dump(ifd, buf, BUFLEN)) > 0) {
		pkt = (struct dpkthdr*) (p = buf);
		rtp = (struct rtphdr*) (buf + DPKTHDRSIZE);
		if (pkt->plen == 0) /* not RTP */
			continue;
		if (verbose)
			print_dpkthdr(pkt);
		if ((hlen = parse_rtphdr(rtp)) == -1) {
			warnx("Error parsing RTP header");
			error = -1;
			continue;
		}
		if (verbose)
			print_rtphdr(rtp);
		if (pkt->dlen - DPKTHDRSIZE < pkt->plen) {
			warnx("%lu bytes of RTP payload missing",
				pkt->plen - pkt->dlen + DPKTHDRSIZE);
		}
		p += DPKTHDRSIZE + hlen;
		r -= DPKTHDRSIZE + hlen;
		if ((w = write(ofd, p, r)) == -1) {
			warnx("Error writing %zd bytes of payload", r);
			error = -1;
			continue;
		} else if (w < r) {
			warnx("Only wrote %zd < %zd bytes of payload", w, r);
			error = -1;
			continue;
		}
	}
	return r == -1 ? -1 : error;
}

int
dump2txt(int ifd, int ofd)
{
	return 0;
}

/* Read RTP packets from the net, write a dump file.
 * Return 0 for success, -1 for error. */
int
net2dump(int ifd, int ofd)
{
	ssize_t r, w;
	int error = 0;
	struct rtphdr *rtp;
	unsigned char buf[BUFLEN];
	struct timeval start;
	if (gettimeofday(&start, NULL) == -1) {
		warn("gettimeofday");
		return -1;
	}
	if (write_dumpline(ofd, addr) == -1) {
		warnx("Error writing dump line");
		return -1;
	}
	if (write_dumphdr(ofd, addr, &start) == -1) {
		warnx("Error writing dump header");
		return -1;
	}
	while ((r = recv(ifd, buf, BUFLEN, 0)) > 0) {
		if (verbose)
			fprintf(stderr, "%zd bytes of RTP received\n", r);
		rtp = (struct rtphdr*) buf;
		if (parse_rtphdr(rtp) == -1) {
			warnx("Error parsing RTP header");
			error = -1;
			continue;
		}
		if (verbose)
			print_rtphdr(rtp);
		if ((w = write_dpkthdr(ofd, r, offset(&start))) == -1) {
			warnx("Error writing dump packet header");
			error = -1;
			continue;
		}
		/* TODO: -s size of RTP to save */
		if ((w = write(ofd, buf, r)) != r) {
			warnx("Error writing %zd bytes of RTP", r);
			error = -1;
			continue;
		}
	}
	return r == -1 ? -1 : error;
}

/* Read RTP packets from the net, write RTP packets to the net.
 * No timing is considered here: write them as you read them.
 * Return 0 for success, -1 for error. */
int
net2net(int ifd, int ofd)
{
	ssize_t s, w;
	int error = 0;
	struct rtphdr *rtp;
	unsigned char buf[BUFLEN];
	while ((s = recv(ifd, buf, BUFLEN, 0)) > 0) {
		if (verbose)
			fprintf(stderr, "%zd bytes of RTP received\n", s);
		rtp = (struct rtphdr*) buf;
		if (parse_rtphdr(rtp) == -1) {
			error = -1;
			warnx("Error parsing RTP header");
			continue;
		}
		if (verbose)
			print_rtphdr(rtp);
		if ((w = write(ofd, buf, s)) == -1) {
			warnx("Error writing %zd bytes of payload", s);
			error = -1;
		} else if (w < s) {
			warnx("Only wrote %zd < %zd bytes of payload", w, s);
			error = -1;
		}
	}
	return s == -1 ? -1 : error;
}

/* Read RTP packets from the ifd, write raw audio payload to ofd.
 * Return 0 for success, -1 for error. */
int
net2raw(int ifd, int ofd)
{
	unsigned char *p;
	unsigned char buf[BUFLEN];
	struct rtphdr *rtp;
	ssize_t s, w, hlen;
	int error = 0;
	while ((s = recv(ifd, buf, BUFLEN, 0)) > 0) {
		if (verbose)
			fprintf(stderr, "%zd bytes of RTP received\n", s);
		rtp = (struct rtphdr*) (p = buf);
		if ((hlen = parse_rtphdr(rtp)) == -1) {
			error = -1;
			warnx("Error parsing RTP header");
			continue;
		}
		if (verbose)
			print_rtphdr(rtp);
		p += hlen;
		s -= hlen;
		if ((w = write(ofd, p, s)) == -1) {
			warnx("Error writing %zd bytes of payload", s);
			error = -1;
		} else if (w < s) {
			warnx("Only wrote %zd < %zd bytes of payload", w, s);
			error = -1;
		}
	}
	return s == -1 ? -1 : error;
}

int
net2txt(int ifd, int ofd)
{
	return 0;
}

int
txt2dump(int ifd, int ofd)
{
	return 0;
}

int
txt2net(int ifd, int ofd)
{
	return 0;
}

int
txt2raw(int ifd, int ofd)
{
	return 0;
}

int
main(int argc, char** argv)
{
	int c;
	int ifd = STDIN_FILENO;
	int ofd = STDOUT_FILENO;

	int (*convert)(int ifd, int ofd) = NULL;
	int (*converter[NUMFORMATS][NUMFORMATS])(int, int) = {
		{ NULL,     dump2net, dump2raw, dump2txt, NULL },
		{ net2dump, net2net,  net2raw,  net2txt,  NULL },
		{ NULL,     NULL,     NULL,	NULL,     NULL },
		{ txt2dump, txt2net,  txt2raw,  NULL,     NULL },
		{ NULL,     NULL,     NULL,     NULL,     NULL },
	};

	while ((c = getopt(argc, argv, "i:o:rtv")) != -1) switch (c) {
		case 'i':
			if (((ifmt = fmtbyname(optarg))) == FORMAT_NONE) {
				warnx("unknown format: %s", optarg);
				return -1;
			}
			break;
		case 'o':
			if ((ofmt = fmtbyname(optarg)) == FORMAT_NONE) {
				warnx("unknown format: %s", optarg);
				return -1;
			}
			break;
		case 'r':
			remote = 1;
			break;
		case 't':
			dumptime = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			usage();
			return -1;
			break;
	}
	argc -= optind;
	argv += optind;

	if (argc > 2) {
		usage();
		return -1;
	}

	if (getifaddrs(&ifaces) == -1)
		err(1, NULL);
	if (-1 == (ifd = (*argv
	? rtpopen(*argv++, O_RDONLY)
	: rtpopen("-",     O_RDONLY)))) {
		warnx("Cannot open input for reading");
		return -1;
	}
	if (-1 == (ofd = (*argv
	? rtpopen(*argv++, O_WRONLY|O_CREAT|O_TRUNC)
	: rtpopen("-",     O_WRONLY|O_CREAT|O_TRUNC)))) {
		warnx("Cannot open output for writing");
		return -1;
	}
	freeifaddrs(ifaces);

	if (ifmt == FORMAT_NONE) {
		warnx("Input format not determined");
		return -1;
	}
	if (ofmt == FORMAT_NONE) {
		warnx("Output format not determined");
		return -1;
	}
	if (ifmt == FORMAT_RAW) {
		warnx("Only output can be raw");
		return -1;
	}
	if ((convert = converter[ifmt][ofmt]) == NULL) {
		warnx("No converter for this input/output combination");
		return -1;
	}
	return convert ? convert(ifd, ofd) : -1;
}
